gtda/externals/python/ripser_interface.py

```diff
@@ -213,29 +213,30 @@ def ripser(X, maxdim=1, thresh=np.inf, coeff=2, metric="euclidean",
         # births
         dm = sparse.coo_matrix(dm)
 
-    if sparse.issparse(dm):
-        if sparse.isspmatrix_coo(dm) and not collapse_edges:
-            # If the matrix is already COO, we need to order the row and column
-            # indices lexicographically to avoid errors. See
-            # https://github.com/scikit-tda/ripser.py/issues/103
-            row, col, data = _lexsort_coo_data(dm.row, dm.col, dm.data)
-        else:
-            # Lexicographic sorting is performed by scipy upon conversion
-            coo = dm.tocoo()
-            row, col, data = coo.row, coo.col, coo.data
+    if sparse.issparse(dm) or collapse_edges:
         if collapse_edges:
-            # Threshold first. Ideally, this would be implemented as a
-            # modification of the C++ code
-            if thresh < np.inf:
-                thresholded_idx = np.flatnonzero(data <= thresh)
-                data = data[thresholded_idx]
-                row = row[thresholded_idx]
-                col = col[thresholded_idx]
-            row, col, data = \
-                gtda_collapser.flag_complex_collapse_edges_from_coo_data_to_coo_data(row, col, data)
-            row, col, data = _lexsort_coo_data(
-                np.asarray(row), np.asarray(col), np.asarray(data)
-                )
+            if not sparse.issparse(dm):
+                row, col, data = \
+                    gtda_collapser.flag_complex_collapse_edges_dense(dm)
+            else:
+                coo = dm.tocoo()
+                row, col, data = \
+                    gtda_collapser.flag_complex_collapse_edges_coo(coo.row,
+                                                                   coo.col,
+                                                                   coo.data)
+        else:
+            if sparse.isspmatrix_coo(dm):
+                # If the matrix is already COO, we need to order the row and
+                # column indices lexicographically to avoid errors. See
+                # https://github.com/scikit-tda/ripser.py/issues/103
+                row, col, data = _lexsort_coo_data(dm.row, dm.col, dm.data)
```

Comment:
With this version, `_lexsort_coo_data` is done twice when the input is coo and `collapse` is `False`. Can we find a way to avoid this?

Reply:
We could replace `row, col, data = _lexsort_coo_data(dm.row, dm.col, dm.data)` by `row, col, data = dm.row, dm.col, dm.data`.

I think it would be the easiest solution

Reply:
Yep, I agree!

Reply:
Given what I wrote above, I think it's enough to just leave the lexsort here and remove it from line 236.

Reply:
But when collapsing edges is enable, we are sure that we shouldn't sort after collapsing ?

Reply:
Oh you are right! So maybe it should be put inside the `if` branch?

Reply:
If I understood you right, only sort if `collapse_edges` is true then ?
Otherwise, when `collapse_edges` is false, it's already sorted because on both cases we have a COO matrix format ?

Reply:
These are the situations in which we need to sort:
- after any collapse (from sparse or dense matrix);
- if the input is COO, regardless of whether we collapse or not.

By the way, I notice that collapse is not yet implemented in the dense case.

Reply:
If I'm not wrong, this means that we should *not* sort when `collapse_edges` is disable *and* matrix is originally in COO format ?

In this case we should add the sorting when `collapse_edges` is enable or if `collapse_edges` is disable *and* matrix format is originally in COO format. 

I would have like to prevent to duplicate code or add a complicated if before calling the sort function, and sort in all cases, even if in the case where the input matrix is originally in sparse format (but not COO) and it's converted to COO format.

What do you think ?

About the dense case, isn't it covered here ? https://github.com/giotto-ai/giotto-tda/pull/469/files#diff-540ff315655f258b36b02c3a50f10946R216-R218

Reply:
Hmm, I'm not sure why you say that, sorry! I said:

> if the input is COO, regardless of whether we collapse or not.

If we don't collapse, other sparse formats are safe. After we collapse, nothing is safe. But if we don't collapse, COO needs the sorting.